lib/charms/kubernetes_charm_libraries/v0/kubernetes_volumes_patch.py

```diff
@@ -0,0 +1,531 @@
+        containers: Iterable[Container] = statefulset.spec.template.spec.containers  # type: ignore[attr-defined]  # noqa: E501
+        requested_volumes_mounts = [
+            requested_volume.volume_mount for requested_volume in requested_volumes
+        ]
+        container = self._get_container(container_name=container_name, containers=containers)
+        container.volumeMounts = [
+            item for item in container.volumeMounts if item not in requested_volumes_mounts
+        ]
+        if self._hugepages_in_requested_volumes(requested_volumes):
+            for limit in HUGEPAGES_RESOURCES_LIMITS.keys():
+                container.resources.limits.pop(limit, None)
+            for request in HUGEPAGES_RESOURCES_REQUESTS.keys():
+                container.resources.requests.pop(request, None)
+        statefulset_volumes = statefulset.spec.template.spec.volumes  # type: ignore[attr-defined]  # noqa: E501
+        requested_volumes_volumes = [
+            requested_volume.volume for requested_volume in requested_volumes
+        ]
+        statefulset.spec.template.spec.volumes = [  # type: ignore[attr-defined]
+            item for item in statefulset_volumes if item not in requested_volumes_volumes
+        ]
+        try:
+            self.client.replace(obj=statefulset)
+        except ApiError:
+            raise KubernetesRequestedVolumesError(
+                f"Could not replace statefulset {statefulset_name}"
+            )
+        logger.info("Requested volumes removed from %s statefulset", statefulset_name)
+
+    @staticmethod
+    def _hugepages_in_requested_volumes(requested_volumes: Iterable[RequestedVolume]) -> bool:
+        """Returns whether requested volumes contain a HugePages volume.
+
+        Args:
+            requested_volumes: Iterable of requested volumes
+
+        Returns:
+            bool: Whether requested volumes contain a HugePages volume
+        """
+        return any(
+            [
+                requested_volume.volume.emptyDir.medium == "HugePages"
+                for requested_volume in requested_volumes
+            ]
+        )
+
+
+class KubernetesVolumesPatchCharmLib(Object):
+    """Class to be instantiated by charms requiring changes in volumes."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        volumes_to_add_func: Callable[[], Iterable[RequestedVolume]],
+        volumes_to_remove_func: Callable[[], Iterable[RequestedVolume]],
+        container_name: str,
+        refresh_event: BoundEvent,
+    ):
+        """Constructor for the KubernetesVolumesPatchLib.
+
+        Args:
+            charm: Charm object
+            volumes_to_add_func: A callable to a function returning a list of
+              `RequestedVolume` to be created.
+            volumes_to_remove_func: A callable to a function returning a list of
+              `RequestedVolume` to be deleted.
```

Comment:
I'm not sure about the charm having to say what it wants to be "added" or "removed". The charm should simply state what should be there and the lib should figure out what to create or delete.

Reply:
+1

Reply:
After re-review, I strongly agree that library should be responsible from identifying the volumes to be created and deleted. This library reaches to Ks8 API and gets what exists there and is it different than the requested one or not. 

From the charm aspect, it validates the requested volume configurations and pass the all the requested volumes details to the library.